gobblin-utility/src/main/java/org/apache/gobblin/util/filesystem/HdfsJarUploadUtils.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.util.filesystem;
+
+import java.io.IOException;
+
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import lombok.extern.slf4j.Slf4j;
+
+
+/**
+ * Utility class for uploading jar files to HDFS with retries to handle concurrency
+ */
+@Slf4j
+public class HdfsJarUploadUtils {
+
+  private static final long WAITING_TIME_ON_IMCOMPLETE_UPLOAD = 3000;
+
+  /**
+   * Calculate the target filePath of the jar file to be copied on HDFS,
+   * given the {@link FileStatus} of a jarFile and the path of directory that contains jar.
+   * Snapshot dirs should not be shared, as different jobs may be using different versions of it.
+   * @param fs
+   * @param localJar
+   * @param unsharedJarsDir
+   * @param jarCacheDir
+   * @return
+   * @throws IOException
+   */
+  public static Path calculateDestJarFile(FileSystem fs, FileStatus localJar, Path unsharedJarsDir, Path jarCacheDir) throws IOException {
+    Path uploadDir = localJar.getPath().getName().contains("SNAPSHOT") ? unsharedJarsDir : jarCacheDir;
+    Path destJarFile = new Path(fs.makeQualified(uploadDir), localJar.getPath().getName());
+    return destJarFile;
+  }
+  /**
+   * Upload a jar file to HDFS with retries to handle already existing jars
+   * @param fs
+   * @param localJar
+   * @param destJarFile
+   * @param jarFileMaximumRetry
+   * @return
+   * @throws IOException
+   */
+  public static boolean uploadJarToHdfs(FileSystem fs, FileStatus localJar, int jarFileMaximumRetry, Path destJarFile) throws IOException {
+    int retryCount = 0;
+    while (!fs.exists(destJarFile) || fs.getFileStatus(destJarFile).getLen() != localJar.getLen()) {
+      try {
+        if (fs.exists(destJarFile) && fs.getFileStatus(destJarFile).getLen() != localJar.getLen()) {
+          Thread.sleep(WAITING_TIME_ON_IMCOMPLETE_UPLOAD);
```

Comment:
what are we waiting on here?  is `copyFromLocalFile` asynchronous?  or are we thinking a different process may be doing upload.

with `overwriteAnyExistingDestFile == false`, we don't seem to undertake any "repair", do we?  instead we just wait `(waitTime * jarFileMaximumRetry)`